lib/srv/app/transport.go

```diff
@@ -0,0 +1,281 @@
+		return nil, trace.Wrap(err)
+	}
+
+	// Forward the request to the target application and emit an audit event.
+	resp, err := t.tr.RoundTrip(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Emit the event to the audit log.
+	if err := t.emitAuditEvent(r, resp); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Perform any response rewriting needed before returning the request.
+	if err := t.rewriteResponse(resp); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return resp, nil
+}
+
+// rewriteRequest applies any rewriting rules to the request before it's forwarded.
+func (t *transport) rewriteRequest(r *http.Request) error {
+	// Update the target address of the request so it's forwarded correctly.
+	r.URL.Scheme = t.uri.Scheme
+	r.URL.Host = t.uri.Host
+
+	// Add in JWT header.
+	r.Header.Add(teleport.AppJWTHeader, t.c.jwt)
+	r.Header.Add(teleport.AppCFHeader, t.c.jwt)
+
+	return nil
+}
+
+// needsPathRedirect checks if the request should be redirected to a different path.
+// At the moment, the only time a redirect happens is if URI specified is not
+// "/" and the public address being requested is "/".
+func (t *transport) needsPathRedirect(r *http.Request) (string, bool) {
+	// If the URI for the application has no path specified, nothing to be done.
+	uriPath := path.Clean(t.uri.Path)
+	if uriPath == "." {
+		uriPath = "/"
+	}
+	if uriPath == "/" {
+		return "", false
+	}
+
+	// For simplicity, only support redirecting to the URI path if the root path
+	// is requested.
+	reqPath := path.Clean(r.URL.Path)
+	if reqPath == "." {
+		reqPath = "/"
+	}
+	if reqPath != "/" {
+		return "", false
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(t.c.publicAddr, t.c.publicPort),
+		Path:   uriPath,
+	}
+	return u.String(), true
+}
+
+// rewriteResponse applies any rewriting rules to the response before returning it.
+func (t *transport) rewriteResponse(resp *http.Response) error {
+	switch {
+	case t.c.rewrite != nil && len(t.c.rewrite.Redirect) > 0:
+		err := t.rewriteRedirect(resp)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+	}
+	return nil
+}
+
+// rewriteRedirect applies redirect rules to the response.
+func (t *transport) rewriteRedirect(resp *http.Response) error {
+	if isRedirect(resp.StatusCode) {
+		// Parse the "Location" header.
+		u, err := url.Parse(resp.Header.Get("Location"))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// If the redirect location is one of the hosts specified in the list of
+		// redirects, rewrite the header.
+		if utils.SliceContainsStr(t.c.rewrite.Redirect, host(u.Host)) {
+			u.Scheme = "https"
+			u.Host = net.JoinHostPort(t.c.publicAddr, t.c.publicPort)
+		}
+		resp.Header.Set("Location", u.String())
+	}
+	return nil
+}
+
+// emitAuditEvent writes the request and response to audit stream.
+func (t *transport) emitAuditEvent(req *http.Request, resp *http.Response) error {
+	appSessionRequestEvent := &events.AppSessionRequest{
```

Comment:
I may be misunderstanding something but these events don't seem to include the session ID? Shouldn't they include it in order to be able to correlate requests/responses from a single session?